server/lib/pbench/indexer.py

```diff
@@ -893,17 +897,294 @@ def get_result_json_dirs(ptb):
+        workload name and version on the first line that we recognize, then we
+        find all the user-benchmark-result.csv files to be indexed.
+        """
+        idxctx = ptb.idxctx
+        ubm_name =  os.path.join(
+                ptb.extracted_root, "user-benchmark-name.txt")
+        try:
+            with open(ubm_name, "r") as fp:
+                first_line = fp.readline()
+                blob = fp.read()
+        except OSError as exc:
+            if exc.errno == errno.ENOENT:
+                return None
+            else:
+                idxctx.logger.warning(
+                        "Unable to read user benchmark name from {}, {}, user"
+                        " benchmark data won't be indexed", ubm_name, exc)
+                return None
+        except Exception as exc:
+            idxctx.logger.error(
+                    "Error reading user benchmark name from {}, {}, user"
+                    " benchmark data won't be indexed", ubm_name, exc)
+            return None
+        user_benchmark = None
+        name, ver = first_line.split(' ', 1)
+        name = name.strip()
+        try:
+            ubm_driver_versions = ResultData.known_user_benchmarks[name]
+        except KeyError:
+            pass
+        else:
+            ver = ver.strip()
+            try:
+                ubm_driver = ubm_driver_versions[ver]
+            except KeyError:
+                idxctx.logger.warning()
+            else:
+                # Verify columns to be safe.
+                cols = set(ubm_driver['columns'].keys())
+                if ubm_driver['val_col'] not in cols:
+                    idxctx.logger.error("Logic bomb! bad ubm driver val_col")
+                    return None
+                if ubm_driver['start_ts_col'] not in cols:
+                    idxctx.logger.error(
+                            "Logic bomb! bad ubm driver start_ts_col")
+                    return None
+                if ubm_driver['end_ts_col'] and (
+                        ubm_driver['end_ts_col'] not in cols):
+                    idxctx.logger.error(
+                            "Logic bomb! bad ubm driver end_ts_col")
+                    return None
+                for bm_md_col in ubm_driver['benchmark_md_cols']:
+                    if bm_md_col not in cols:
+                        idxctx.logger.error(
+                                "Logic bomb! bad ubm driver benchmark_md_cols")
+                        return None
+                for sample_md_col in ubm_driver['sample_md_cols']:
+                    if sample_md_col not in cols:
+                        idxctx.logger.error(
+                                "Logic bomb! bad ubm driver sample_md_cols")
+                        return None
+                _dict_const(
+                    name=name, version=ver.strip(),
+                    description=blob.strip(),
+                    driver=ubm_driver)
+        return user_benchmark
```

Comment:
Few Suggestions: 
1. According to the docString, this function is used to determine if this is a known user benchmark so it can be more readable if it returns boolean.
2. Wouldn't this function always return `None` because user_benchmark is set to None (on line 979) but never gets reassigned?
3. I think there are a lot of if statements with similar logger messages and they all return None so it can be rewritten to be more concise. Maybe `set` operations can be useful something like
```
cols = set(ubm_driver['columns'].keys())
cols_to_check = set(list(ubm_driver['val_col']) + list(ubm_driver['start_ts_col']) + list(ubm_driver['end_ts_col']) + ubm_driver['benchmark_md_cols'] + ubm_driver['sample_md_cols'] )
set_diff = cols_to_check - cols  # since we want to check if all the items in cols_to_check present in cols
if len(set_diff) >= 1:
  idxctx.logger.error("Logic bomb! bad ubm driver: {}".format(' '.join(map(str, set_diff))))
  return None
```
The above code may not log the field names such as `val_col ` or `start_ts_col` etc. but the logged column names can be traced back to these fields.

Reply:
Reworked this code.